keps/sig-api-machinery/2211-strip-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,426 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+ManagedFields is used to keep track of who has changed a field of an object, and Server-side Apply relies on such information.
+While the managedFields is beneficial to resolving conflicts on updates, it is a bit verbose to many users because this field
+could be quite large and generally users are not interested in this information.
+
+This KEP is proposing stripping the managedFields on the server-side when users request objects to make the response less verbose.
+
+## Motivation
+
+Many users have complained in [kubernetes#90066][90066] that they got impacted by the addition of managedFields
+(and the issue itself has received more than 100 thumb-up reactions), and we want to improve the user experience.
+
+[90066]: https://github.com/kubernetes/kubernetes/issues/90066
+
+At the same time, it is not ideal to only drop the managedFields on the client-side, because that would require every client
+(such as kubectl or CLI tools written by users using client-go) to do the same thing.
+
+With this KEP, the API server could return objects without managedFields on demand, so that least client-side changes is needed
+to make the response less verbose. 
+
+### Goals
+
+* Strip managedFields on the server-side.
+
+### Non-Goals
+
+* Turn off managedFields. Server-side Apply relies on managedFields, we could not simply disable it.
+* Clear manageFields only on the client-side.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Server-side changes
+
+* Add three content types to denote the client expects the API server to strip the managedFields.
+  * `application/json+unmanaged`
+  * `application/yaml+unmanaged`
+  * `application/vnd.kubernetes.protobuf+unmanaged`
+* Add a new serializer, say `UnmanagedSerializer`, which is able to strip the managedFields when encoding objects.
+* Make the API server to use `UnmanagedSerializer` when encounters the above three content type.
+
+### Client-side changes
+
+FIXME: we need to set the `Accept` header on the client-side and perform client-side stripping if the API server still
+returns objects with managedFields. I am not sure we should do this in `client-go` and `cli-runtime` or only in `kubectl get`.
+
+/cc @apelisse
```

Comment:
/cc @apelisse

Reply:
what about controller-runtime, or is that taken care of by the client-go ?

Reply:
My proposed change is providing a helper function in `client-go` which is able to set the `Accept` header in `rest.Config`, so that users could tell the API server they want to omit the managed fields.

Then you could use the modified `rest.Config` everywhere that needs it and the managed fields should be dropped if the API server recognize the new `Accept` header.

Reply:
I'm still really not convinced we need a mechanism to set a single Accept header.

Reply:
I wouldn't say `need` but the header we propose for fallback etc. might not be too known or familiar as an option here that I wouldn't mind having a simpler option that does it for you.
Like `PreferNoManagedFields()` = header with `q=` fallback and `RequestNoManagedFields()` = no fallback.

Reply:
@kwiesmueller yes that's what I mean by saying "providing a helper function", but it seems there is more we could do on client-side if we need to consider the case that API server cannot recognize this new header and we still want to omit the managed fields.